### ISSUE TITLE
Unable to restore Cache

### ISSUE THREAD

**Issue by nascit** [actions/cache#1052]
I'm on GHES 3.7.2 using a standalone Amazon Linux 2 instance as my self-hosted runner. When trying to recover a cache, it seems I'm getting a file different from the one I have uploaded (Cache Size: ~0 MB (423 B) ).

I'm creating a simple file to cache to make the example cleaner. Workflow used:

```
jobs:
  cache:
    name: Cache artifact
    runs-on: self-hosted
    steps:
    - uses: actions/checkout@v3

    - name: Set up JDK
      uses: actions/setup-java@v3
      with:
        java-version: '11'
        distribution: 'temurin

    - name: Cache Maven packages
      uses: actions/cache@v3
      env:
        cache-name: cache-maven
        SEGMENT_DOWNLOAD_TIMEOUT_MINS: 30
      with:
        path: |
          ${{ github.workspace }}/.m2/repository
        key: ${{ runner.os }}-build-${{ env.cache-name }}-${{ hashFiles('**/pom.xml') }}
        restore-keys: |
            ${{ runner.os }}-build-${{ env.cache-name }}-
            ${{ runner.os }}-build-
            ${{ runner.os }}-

    - name: Cache hit
      if: steps.cache.outputs.cache-hit != 'true'
      run: echo "Cache missed"
    
    - name: Populate m2 to mock Maven build
      run: |
        mkdir -pv ${{ github.workspace }}/.m2/repository/
        echo "helloooooooooooo" > ${{ github.workspace }}/.m2/repository/hello.txt
```

As expected, after first run, the cache is saved:

```
##[debug]Uploading chunk of size 238 bytes at offset 0 with content range: bytes 0-237/*
##[debug]Commiting cache
Cache Size: ~0 MB (238 B)
##[debug]Resource Url: https://github.domain.com/_services/artifactcache/cveBwaOnecQqdkvCBH5kKVWeBdERBEWwsIreLsV53YNAHZLkUZ/_apis/artifactcache/caches/35
Cache saved successfully
Cache saved with key: Linux-build-cache-maven-016ccdc03fcd75fdb5440854e86c2f1833f3d142f4aad77779c983999d3c8585
```

Using GH CLI to query the caches I can see it exists:
```
> gh actions-cache list -B main

Showing 1 of 1 cache entries in ORG/java-supply-chain

Linux-build-cache-maven-016ccdc03fcd75fdb5440854e86c2f1833f3d142f4aad77779c983999d3c8585  238.00 B  refs/heads/main  4 minutes ago
```

Second time it runs, I can see by enabling DEBUG logs that the cache seems to be found (even gives me the URL), but still it couldn't be downloaded:

```
##[debug]Loading env
Run actions/cache@v3
  with:
    path: /runner/_work/workspace/workspace/.m2/repository
    key: Linux-build-cache-maven-016ccdc03fcd75fdb5440854e86c2f1833f3d142f4aad77779c983999d3c8585
    restore-keys: Linux-build-cache-maven-
  Linux-build-
  Linux-
  
  env:
    JAVA_HOME: /opt/hostedtoolcache/Java_Temurin-Hotspot_jdk/11.0.17-8/x64
    cache-name: cache-maven
    SEGMENT_DOWNLOAD_TIMEOUT_MINS: 30
##[debug]Resolved Keys:
##[debug]["Linux-build-cache-maven-016ccdc03fcd75fdb5440854e86c2f1833f3d142f4aad77779c983999d3c8585","Linux-build-cache-maven-","Linux-build-","Linux-"]
##[debug]Checking zstd --version
##[debug]*** zstd command line interface 64-bits v1.4.4, by Yann Collet ***
##[debug]Resource Url: https://github.domain.com/_services/artifactcache/cveBwaOnecQqdkvCBH5kKVWeBdERBEWwsIreLsV53YNAHZLkUZ/_apis/artifactcache/cache?keys=Linux-build-cache-maven-016ccdc03fcd75fdb5440854e86c2f1833f3d142f4aad77779c983999d3c8585%2CLinux-build-cache-maven-%2CLinux-build-%2CLinux-&version=3fb9331ccdb11fe063e37df363ce34b0a4ce8cf0b743d09dc30722ac42cae114
::add-mask::***
##[debug]Cache Result:
##[debug]{"scope":"refs/heads/main","cacheKey":"Linux-build-cache-maven-016ccdc03fcd75fdb5440854e86c2f1833f3d142f4aad77779c983999d3c8585","cacheVersion":"3fb9331ccdb11fe063e37df363ce34b0a4ce8cf0b743d09dc30722ac42cae114","creationTime":"2023-01-04T17:34:14.7166667Z","archiveLocation":"***"}
##[debug]Archive Path: /runner/_work/_temp/3a356310-7a04-4d0d-a8bd-9bf88da823d3/cache.tzst
##[debug]Use Azure SDK: true
##[debug]Download concurrency: 8
##[debug]Request timeout (ms): 30000
##[debug]Cache segment download timeout mins env var: 30
##[debug]Segment download timeout (ms): 1800000
##[debug]Unable to validate download, no Content-Length header
/usr/bin/tar --use-compress-program unzstd -tf /runner/_work/_temp/3a356310-7a04-4d0d-a8bd-9bf88da823d3/cache.tzst -P
zstd: /*stdin*\: unsupported format 
/usr/bin/tar: Child returned status 1
/usr/bin/tar: Error is not recoverable: exiting now
Warning: Failed to restore: Tar failed with error: The process '/usr/bin/tar' failed with exit code 2
Cache not found for input keys: Linux-build-cache-maven-016ccdc03fcd75fdb5440854e86c2f1833f3d142f4aad77779c983999d3c8585, Linux-build-cache-maven-, Linux-build-, Linux-
##[debug]Node Action run completed with exit code 0
##[debug]Save intra-action state CACHE_KEY = Linux-build-cache-maven-016ccdc03fcd75fdb5440854e86c2f1833f3d142f4aad77779c983999d3c8585
##[debug]Finishing: Cache Maven packages
```

If I disable Debug logs I have slightly different logs:

```
Run actions/cache@v3
Cache Size: ~0 MB (423 B)
/usr/bin/tar --use-compress-program unzstd -xf /runner/_work/_temp/098142-5211-41e2-9727-501837e79272/cache.tzst -P -C /runner/_work/java-supply-chain/java-supply-chain
zstd: /*stdin*\: unsupported format 
/usr/bin/tar: Child returned status 1
/usr/bin/tar: Error is not recoverable: exiting now
Warning: Failed to restore: Tar failed with error: The process '/usr/bin/tar' failed with exit code 2
Cache not found for input keys: Linux-build-cache-maven-016ccdc03fcd75fdb5440854e86c2f1833f3d142f4aad77779c983999d3c8585, Linux-build-cache-maven-, Linux-build-, Linux-
```

Cache size should be 238 B, and when trying to download it shows 423 B, so maybe I'm getting an unexpected file instead?
Any idea how can I further troublshoot this issue?

**Comment by tanuj077** [comment]
Hi @nascit, we are looking into this, if immediate assistance is required please create a support ticket with github support.

**Comment by pallavx** [comment]
Hi @nascit, for our investigation, can you please provide the complete log for the workflow? It would also help if you provide a list of all caches.

**Comment by nascit** [comment]
@pallavx I have updated the issue with more logs and a better explanation

**Comment by pallavx** [comment]
Hi @nascit,

In your `name: Cache Maven packages` step, you need to assign an `id` to the step, so that the output can be checked in the next step `if: steps.cache.outputs.cache-hit != 'true'`.

For example, if you add `id` to the `name: Cache Maven packages` step, it would work.

```yaml
name: Cache Maven packages
uses: actions/cache@v3
id: cache
```

Reference: https://docs.github.com/en/actions/using-jobs/defining-outputs-for-jobs

**Comment by nascit** [comment]
Hey @pallavx 
We got in contact with GHES support.
This is error was actually due to GitHub not using Sigv4 when using S3 as the backend for cache artifacts, and only occurs when customer managed KMS is used.
Hopefully it will be fixed in GHES 3.8
